Hide __getattr__ from type checkers

hauntsaninja opened the pull request:
Visibility of this causes type checkers to conservatively assume that all attributes are defined on torch module.

cc @ezyang @malfet @rgommers @xuzhao9 @gramster

hauntsaninja committed:
Hide __getattr__ from type checkers

Visibility of this causes type checkers to conservatively assume that
all attributes are defined on torch module.

hauntsaninja commented:
@pytorchbot label "module: typing"

hauntsaninja commented:
@pytorchbot label "release notes: python_frontend"

hauntsaninja committed:
add missing dtypes

malfet commented on torch/__init__.py:
```diff
@@ -1775,20 +1775,22 @@ def registerOp(cls, op_key, full_schema, op_impl, dispatch_key):
+            import importlib
+            return importlib.import_module(f".{name}", __name__)
+
+        raise AttributeError(f"module '{__name__}' has no attribute '{name}'")
```
It would be nice to merge it with `if TYPE_CHECKING:` condition in `__init__.py` so that it would like like
```python
if TYPE_CHECKING:
  import torch.onnx
  import torch.compile
else:
 def __getattr__:
   ...

hauntsaninja replied:
Thanks, updated!

hauntsaninja commented:
@pytorchbot merge

hauntsaninja committed:
code review

hauntsaninja committed:
Merge remote-tracking branch 'upstream/main' into getattr-typecheck

malfet commented:
@pytorchbot merge

hauntsaninja committed:
maybe fix some

hauntsaninja committed:
add things

hauntsaninja committed:
type ignore the rest

hauntsaninja commented:
Rebased! I fixed some of the type errors that got introduced and hidden after my previous merge base, but just type ignored others. It would be nice to merge before too many further regressions are committed to main.

hauntsaninja committed:
shadow

hauntsaninja committed:
ignore like the others do

ezyang commented:
@pytorchbot merge
